src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -246,5 +251,23 @@ public void GetResult()
                 throw new IOException(error.Message, error);
             }
         }
+
+        public void Dispose()
+        {
+            AbortAwaiting();
+
+            // Return all blocks
+            var block = _head;
```

Comment:
This is only called after the socket is closed so there should not be any more incoming data. The AbortAwaiting call kills any consumers. (If all else fails, the finalizer still returns the block in production, but it would crash the process in debug builds)